### PR TITLE
Add checkpoints

### PR DESCRIPTION
**Pull request by dh-ecc**
This PR addresses #326. It is currently working, though there are a couple of open questions / TODOs that would help with keeping the code easier to maintain. They are not necessarily blockers, and can be deferred / worked around. I have grouped these into 2 lists below TODO and TODO-alt respectively.

TODO:
- [x] ~Decide on the format of the json for the checkpoint files, specifically should the value for `height` be a `number` or a `string`, and what we expect the precision to be. There is currently a disparity between what `zcashd` produces from the `z_gettreestate` rpc call and what `gettreestate` from `lightwalletd` produces (the former uses a number in the json response for `height`, the latter uses a string). Depending on our decision this may also involve an update to `lightwalletd`. This is on the assumption that we want the checkpoint files to exactly match the output of `lightwalletd`.~ **Going with a `string`**
- [x] ~Update the checkpoint files in the android and iOS SDKs to reflect this.~ **tracked as separate Issue in https://github.com/zcash/zcash-android-wallet-sdk/issues/442)**
- [x] ~Align the rust SDK with the decisions above.~ **Deferred to a later date**

TODO-alt:
- [ ] ~Decide on the format of the json for the checkpoint files, specifically should the value for `height` be a `number` or a `string`. _however_ we could make this decision independant of the output of `lightwalletd`. OR defer this decision entirely and assume that it can be _either_ a number or a string.~
- [ ] ~Leave the checkpoint files as they are~
- [ ] ~defer the aligning the rust library and `zcashd` / `lightwalletd` int representations (for block height).~

--- 

This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md).

# Author

- [x] Self-review: Did you review your own code in GitHub's web interface? Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs.
- [x] Automated tests: Did you add appropriate automated tests for any code changes?
- [ ] Code coverage: Did you check the code coverage report for the automated tests?  While we are not looking for perfect coverage, the tool can point out potential cases that have been missed.
- [ ] Documentation: Did you update Docs as appropiate? (E.g [README.md](../blob/main/README.md), etc.)
- [ ] Run the app: Did you run the app and try the changes? 
- [ ] Did you provide Screenshots of what the App looks like before and after your changes as part of the description of this PR? (only applicable to UI Changes)
- [ ] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit.

# Reviewer

- [x] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md) checklist?
- [x] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [x] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?_You will find manual testing guidelines under our [manual testing section](../blob/mater/docs/testing/manual_testing)_
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage befor and after your changes and when possible, leave it in a better place. [Learn More...](../blob/master/docs/testing/local_coverage.md)_
- [ ] Documentation: Did you review Docs, [README.md](../blob/master/README.md), [LICENSE.md](../blob/master/LICENSE.md), and [Architecture.md](../blob/master/docs/Architecture.md) as appropriate?
- [x] Run the app: Did you run the app and try the changes? While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data.

### CODE REVIEW

File: Sources/ZcashLightClientKit/Constants/WalletBirthday+Constants.swift [hunk 1]
```diff
@@ -17,3 +17,60 @@ public extension WalletBirthday {
...
+        tree = try container.decode(String.self, forKey: .tree)
+    }
+
+    static func getHeight(from container: KeyedDecodingContainer<CodingKeys>) throws -> Int {
+        if let heightString = try? container.decode(String.self, forKey: .height),
+           let height = Int(heightString) {
+            return height
+        } else {
+            return try container.decode(Int.self, forKey: .height)
+        }
```
[the remark this answers is not in the document]
Comment (LukasKorba) [reply Sources/ZcashLightClientKit/Constants/WalletBirthday+Constants.swift:74]: @adamstener btw, there's double space in the else { body

**Comment by pacu** [comment]
> Decide on the format of the json for the checkpoint files, specifically should the value for height be a number or a string, and what we expect the precision to be. There is currently a disparity between what zcashd produces from the z_gettreestate rpc call and what gettreestate from lightwalletd produces (the former uses a number in the json response for height, the latter uses a string). Depending on our decision this may also involve an update to lightwalletd. This is on the assumption that we want the checkpoint files to exactly match the output of lightwalletd.

@dh-ecc I think the right way to go is to match android's implementation, then take it from there.

**Comment by pacu** [comment]
Note: I marked a few checkpoints with a warning. then I realized there were thousands of them 😂

File: Sources/ZcashLightClientKit/Constants/WalletBirthday+Constants.swift [hunk 2]
```diff
@@ -17,3 +17,60 @@ public extension WalletBirthday {
         }
     }
 }
+
+extension WalletBirthday {
+    static func checkpointsFromDirectory(at url: URL) -> [WalletBirthday] {
```
[the remark this answers is not in the document]
Comment (dh-ecc) [reply Sources/ZcashLightClientKit/Constants/WalletBirthday+Constants.swift:22]: > From what I see this method loads every file into memory. Which shouldn't be the case.

Yeh it is for the moment 😅. I will update the implementation, I stopped at the "make it work" stage, but I'll give it a "make it fast" rewrite shortly 😄 .

> there are two use cases:
>   1. get newest checkpoint.
>   2. get the closest checkpoint to height H rounding down until `SAPLING_ACTIVATION`.

That's interesting, I've never considered `1.` seperately! I've seen as a special case of `2.` where H is higher than our newest checkpoint. Is there a context where we want this functionality in the form of `1.` explicitely?

> on the first case [...]

> on the second case [...]

👍 Sounds good, that's the intended implementation.

[the remark this answers is not in the document]
Comment (pacu) [reply Sources/ZcashLightClientKit/Constants/WalletBirthday+Constants.swift:22]: >Is there a context where we want this functionality in the form of 1. explicitely?

For the case of "create new wallet" you already know that the only checkpoint. you are interested in is the latest/newest/more recent you have in the bundle.

**Review by pacu — CHANGES_REQUESTED** [review]
LGTM! There's a small change I want to discuss in terms of making that birthday file search more efficient.

**Comment by pacu** [comment]
Looks good to me.

I think we are missing declaring the resources on the `podspec` file. 
Also swift-grpc might need to be updated to a newer version to match the one used by SPM

**Review by pacu — CHANGES_REQUESTED** [review]
Looks good to me.

I think we are missing declaring the resources on the `podspec` file. 
Also swift-grpc might need to be updated to a newer version to match the one used by SPM

**Comment by dh-ecc** [comment]
> I think we are missing declaring the resources on the `podspec` file.
– @pacu 

💯 good catch! Adding this now.

**Review by pacu — APPROVED** [review]

**Review by pacu — APPROVED** [review]

**Pull request merged**